File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/rendering/components/A2UIText.kt
package com.interfaceai.a2ui.rendering.components

import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.sp
import com.interfaceai.a2ui.models.A2UIComponent
import com.interfaceai.a2ui.processing.DataBindingResolver
import com.interfaceai.a2ui.processing.TokenResolver
import com.interfaceai.a2ui.rendering.applyA2UIStyle

@Composable
fun A2UITextView(
    component: A2UIComponent,
    tokenResolver: TokenResolver,
    dataResolver: DataBindingResolver,
    modifier: Modifier = Modifier
) {
    val resolvedText = dataResolver.resolve(component.text)
    val style = component.style

    val fontSize = tokenResolver.resolveStyleFloat(style, "fontSize")
    val fontWeight = tokenResolver.resolveStyleString(style, "fontWeight")
    val color = tokenResolver.resolveColor(style, "color")
    val textAlign = tokenResolver.resolveStyleString(style, "textAlign")

    Text(
        text = resolvedText,
        modifier = modifier.applyA2UIStyle(style, tokenResolver),
        fontSize = fontSize?.sp ?: 14.sp,
        fontWeight = mapFontWeight(fontWeight),
        color = color ?: androidx.compose.ui.graphics.Color.Unspecified,
        textAlign = mapTextAlign(textAlign)
    )
}

private fun mapFontWeight(weight: String?): FontWeight {
    if (weight == null) return FontWeight.Normal
    val w = weight.lowercase()
    return when {
        w.contains("bold") -> FontWeight.Bold
        w.contains("semibold") || w.contains("semi") -> FontWeight.SemiBold
        w.contains("medium") -> FontWeight.Medium
        w.contains("light") -> FontWeight.Light
        w.contains("thin") -> FontWeight.Thin
        else -> FontWeight.Normal
    }
}

private fun mapTextAlign(align: String?): TextAlign? {
    return when (align?.lowercase()) {
        "center" -> TextAlign.Center
        "right" -> TextAlign.End
        "left" -> TextAlign.Start
        else -> null
    }
}

File: ios-renderer/Sources/A2UIRenderer/Rendering/Components/A2UIText.swift
import SwiftUI

/// Renders an A2UI "Text" component as a SwiftUI Text view.
public struct A2UITextView: View {
    let component: A2UIComponent
    let tokenResolver: TokenResolver
    let dataResolver: DataBindingResolver

    public var body: some View {
        let resolvedText = dataResolver.resolve(component.text)
        let style = component.style

        let fontSize = tokenResolver.resolveStyleCGFloat(style, key: "fontSize")
        let fontWeight = tokenResolver.resolveStyleString(style, key: "fontWeight")
        let fontFamily = tokenResolver.resolveStyleString(style, key: "fontFamily")
        let color = tokenResolver.resolveColor(style, key: "color")
        let textAlign = tokenResolver.resolveStyleString(style, key: "textAlign")
        let lineHeight = tokenResolver.resolveStyleCGFloat(style, key: "lineHeight")
        let letterSpacing = tokenResolver.resolveStyleCGFloat(style, key: "letterSpacing")

        Text(resolvedText)
            .if(fontFamily != nil && fontSize != nil) {
                $0.font(.custom(fontFamily!, size: fontSize!))
            }
            .if(fontFamily == nil && fontSize != nil) {
                $0.font(.system(size: fontSize!))
            }
            .if(fontWeight != nil) { $0.fontWeight(mapFontWeight(fontWeight!)) }
            .if(color != nil) { $0.foregroundColor(color!) }
            .if(textAlign != nil) { $0.multilineTextAlignment(mapTextAlignment(textAlign!)) }
            .if(lineHeight != nil && fontSize != nil) {
                $0.lineSpacing(lineHeight! - (fontSize ?? 14))
            }
            .if(letterSpacing != nil) { $0.tracking(letterSpacing!) }
            .applyA2UIStyle(component.style, tokenResolver: tokenResolver)
    }
}

private func mapTextAlignment(_ align: String) -> TextAlignment {
    switch align.lowercased() {
    case "center": return .center
    case "right": return .trailing
    default: return .leading
    }
}
